### ISSUE TITLE
Can't Vagrant up if npm_packages configured

### ISSUE THREAD

**Issue by tlartaud** [puphpet/puphpet#1895]
Hi there.

Running:
Windows 10
VirtualBox 5.0.4
Vagrant 1.7.4

`vagrant up` run successfully if I have in my config.yaml file
```
nodejs:
    install: '1'
    npm_packages: {  }
```

However, if I had some packages like the following
```
nodejs:
    install: '1'
    npm_packages:
        - bower
        - grunt-cli
```

I receive the following error
```
==> default: Error: Command npm is missing
==> default: Error: /Stage[main]/Puphpet_nodejs/Package[grunt-cli]/ensure: change from absent to present failed: Command npm is missing
```

Here is the complete config.yaml
```
vagrantfile:
    target: local
    vm:
        box: puphpet/ubuntu1404-x64
        box_url: puphpet/ubuntu1404-x64
        hostname: vm.myproject.dev
        memory: '2048'
        cpus: '4'
        chosen_provider: virtualbox
        network:
            private_network: 192.168.56.108
            forwarded_port:
                vflnp_ws1o17zbvarf:
                    host: '5041'
                    guest: '22'
        post_up_message: ''
        provider:
            virtualbox:
                modifyvm:
                    natdnshostresolver1: 'on'
                showgui: ''
            vmware:
                numvcpus: 1
            parallels:
                cpus: 1
        provision:
            puppet:
                manifests_path: puphpet/puppet
                manifest_file: site.pp
                module_path: puphpet/puppet/modules
                options:
                    - '--verbose'
                    - '--hiera_config /vagrant/puphpet/puppet/hiera.yaml'
                    - '--parser future'
        synced_folder:
            vflsf_nc5ox9jgmng3:
                source: ./
                target: /var/www
                sync_type: default
                smb:
                    smb_host: ''
                    smb_username: ''
                    smb_password: ''
                rsync:
                    args:
                        - '--verbose'
                        - '--archive'
                        - '-z'
                    exclude:
                        - .vagrant/
                        - .git/
                    auto: 'true'
                owner: www-data
                group: www-data
        usable_port_range:
            start: 10200
            stop: 10500
    ssh:
        host: null
        port: null
        private_key_path: null
        username: vagrant
        guest_port: null
        keep_alive: true
        forward_agent: false
        forward_x11: false
        shell: 'bash -l'
    vagrant:
        host: detect
server:
    install: '1'
    packages:
        - vim
        - curl
        - build-essential
        - git-core
        - php-console-table
        - imagemagick
        - sendmail
        - unzip
        - mc
        - graphviz
        - htop
users_groups:
    install: '1'
    groups: {  }
    users: {  }
locale:
    install: '1'
    settings:
        default_locale: 'fr_FR.UTF-8'
        locales:
            - fr_FR.UTF-8
            - en_GB.UTF-8
            - en_US.UTF-8
firewall:
    install: '1'
    rules: {  }
cron:
    install: '1'
    jobs: {  }
nginx:
    install: '0'
    settings:
        default_vhost: 1
        proxy_buffer_size: 128k
        proxy_buffers: '4 256k'
    upstreams: {  }
    vhosts:
        nxv_2q8awr7wxkrw:
            server_name: awesome.dev
            server_aliases:
                - www.awesome.dev
            www_root: /var/www/awesome
            listen_port: '80'
            index_files:
                - index.html
                - index.htm
                - index.php
            client_max_body_size: 1m
            ssl: '0'
            ssl_cert: ''
            ssl_key: ''
            ssl_port: '443'
            ssl_protocols: ''
            ssl_ciphers: ''
            rewrite_to_https: '1'
            spdy: '1'
            locations:
                nxvl_oib9bgms0ggv:
                    location: /
                    autoindex: 'off'
                    internal: 'off'
                    try_files:
                        - $uri
                        - $uri/
                        - /index.php$is_args$args
                    fastcgi: ''
                    fastcgi_index: ''
                    fastcgi_split_path: ''
                nxvl_scbx884b0p8d:
                    location: '~ \.php$'
                    autoindex: 'off'
                    internal: 'off'
                    try_files:
                        - $uri
                        - $uri/
                        - /index.php$is_args$args
                    fastcgi: '127.0.0.1:9000'
                    fastcgi_index: index.php
                    fastcgi_split_path: '^(.+\.php)(/.*)$'
                    fast_cgi_params_extra:
                        - 'SCRIPT_FILENAME $request_filename'
                        - 'APP_ENV dev'
    proxies: {  }
apache:
    install: '1'
    settings:
        user: www-data
        group: www-data
        default_vhost: true
        manage_user: false
        manage_group: false
        sendfile: 0
    modules:
        - proxy_fcgi
        - rewrite
    vhosts:
        av_sebhmghpjrv7:
            servername: myproject.dev
            serveraliases:
                - www.myproject.dev
            docroot: /var/www/public
            port: '80'
            setenv:
                - 'APP_ENV dev'
                - 'NODE_ENV development'
            custom_fragment: ''
            ssl: '0'
            ssl_cert: ''
            ssl_key: ''
            ssl_chain: ''
            ssl_certs_dir: ''
            ssl_protocol: ''
            ssl_cipher: ''
            directories:
                avd_sigtc4yb410o:
                    path: /var/www/public
                    options:
                        - Indexes
                        - FollowSymlinks
                        - MultiViews
                    allow_override:
                        - All
                    require:
                        - 'all granted'
                    custom_fragment: ''
                    files_match:
                        avdfm_e63x92ndinf5:
                            path: \.php$
                            sethandler: 'proxy:fcgi://127.0.0.1:9000'
                            custom_fragment: ''
                            provider: filesmatch
                    provider: directory
php:
    install: '1'
    settings:
        version: '56'
    modules:
        php:
            - cli
            - intl
            - mcrypt
        pear: {  }
        pecl:
            - pecl_http
    ini:
        display_errors: 'On'
        error_reporting: '-1'
        session.save_path: /var/lib/php/session
        date.timezone: UTC
    fpm_ini:
        error_log: /var/log/php-fpm.log
    fpm_pools:
        phpfp_yo2zo1b9pl3r:
            ini:
                prefix: www
                listen: '127.0.0.1:9000'
                security.limit_extensions: .php
                user: www-user
                group: www-data
    composer: '1'
    composer_home: ''
xdebug:
    install: '1'
    settings:
        xdebug.default_enable: '1'
        xdebug.remote_autostart: '0'
        xdebug.remote_connect_back: '1'
        xdebug.remote_enable: '1'
        xdebug.remote_handler: dbgp
        xdebug.remote_port: '9000'
blackfire:
    install: '0'
    settings:
        server_id: ''
        server_token: ''
        agent:
            http_proxy: ''
            https_proxy: ''
            log_file: stderr
            log_level: '1'
        php:
            agent_timeout: '0.25'
            log_file: ''
            log_level: '1'
xhprof:
    install: '0'
wpcli:
    install: '1'
    version: v0.19.0
drush:
    install: '0'
    version: 6.3.0
ruby:
    install: '1'
    versions: {  }
python:
    install: '1'
    packages: {  }
    versions: {  }
nodejs:
    install: '1'
    npm_packages: {  }
hhvm:
    install: '0'
    nightly: 0
    composer: '1'
    composer_home: ''
    settings: {  }
    server_ini:
        hhvm.server.host: 127.0.0.1
        hhvm.server.port: '9000'
        hhvm.log.use_log_file: '1'
        hhvm.log.file: /var/log/hhvm/error.log
    php_ini:
        display_errors: 'On'
        error_reporting: '-1'
        date.timezone: UTC
mysql:
    install: '1'
    settings:
        version: '5.6'
        root_password: '123'
        override_options: {  }
    adminer: 0
    users:
        mysqlnu_0plcds180kpx:
            name: wp
            password: wp
    databases:
        mysqlnd_7ct0d0wk99tk:
            name: myproject
            sql: ''
    grants:
        mysqlng_cmweo1ph3ebi:
            user: wp
            table: '*.*'
            privileges:
                - ALL
mariadb:
    install: '0'
    settings:
        version: '10.0'
        root_password: '123'
        override_options: {  }
    adminer: 0
    users:
        mariadbnu_1w0gwysmrjgp:
            name: dbuser
            password: '123'
    databases:
        mariadbnd_7msz4f4xd7vb:
            name: dbname
            sql: ''
    grants:
        mariadbng_vzgzwcddo9l1:
            user: dbuser
            table: '*.*'
            privileges:
                - ALL
postgresql:
    install: '0'
    settings:
        global:
            encoding: UTF8
            version: '9.3'
        server:
            postgres_password: '123'
    databases: {  }
    users: {  }
    grants: {  }
    adminer: 0
mongodb:
    install: '0'
    settings:
        auth: 1
        bind_ip: 127.0.0.1
        port: '27017'
    databases: {  }
redis:
    install: '0'
    settings:
        conf_port: '6379'
sqlite:
    install: '0'
    adminer: 0
    databases: {  }
mailcatcher:
    install: '1'
    settings:
        smtp_ip: 0.0.0.0
        smtp_port: 1025
        http_ip: 0.0.0.0
        http_port: '1080'
        mailcatcher_path: /usr/local/rvm/wrappers/default
        from_email_method: inline
beanstalkd:
    install: '0'
    settings:
        listenaddress: 0.0.0.0
        listenport: '13000'
        maxjobsize: '65535'
        maxconnections: '1024'
        binlogdir: /var/lib/beanstalkd/binlog
        binlogfsync: null
        binlogsize: '10485760'
    beanstalk_console: 0
rabbitmq:
    install: '0'
    settings:
        port: '5672'
    users: {  }
    vhosts: {  }
    plugins: {  }
elastic_search:
    install: '0'
    settings:
        version: 1.4.1
        java_install: true
solr:
    install: '0'
    settings:
        version: 4.10.2
        port: '8984'
```

Is there any configuration I am missing ?
Thank you for your help.

**Comment by kusmierz** [comment]
Important notices from the main page of puphpet:
* VirtualBox 5.x is causing issues with Vagrant. Make sure you are on 4.3.x!
* Windows 10 hosts are not yet supported!

**Comment by tlartaud** [comment]
Damn,

On the VBox homepage, the notice about Windows 10 that wasn't officially supported disapeared, so i switched to 5.x

I will roll back to version 4.3.30 and let you know.

Thanks for the quick answer.

**Comment by tlartaud** [comment]
@kusmizer @jtreminio The main problem is that Virtual 4.3.30 fail really hard on W10.

Got thoses errors messages, need to vagrant uup 4 times to get the VM finally booted.
```
00:00:00.978374 SUP: Loaded VBoxDDR0.r0 (C:\Program Files\Oracle\VirtualBox\VBoxDDR0.r0) at 0xfffff8002e160000 - ModuleInit at 0000000000000000 and ModuleTerm at 0000000000000000 using the native ring-0 loader
00:00:00.978382 SUP: windbg> .reload /f C:\Program Files\Oracle\VirtualBox\VBoxDDR0.r0=0xfffff8002e160000
00:00:02.498833 SUP: Loading failed for VBoxDD2R0.r0 (C:\Program Files\Oracle\VirtualBox\VBoxDD2R0.r0) rc=VERR_LDR_MISMATCH_NATIVE
00:00:02.498940 pdmR3LoadR0U: pszName="VBoxDD2R0.r0" rc=VERR_LDR_MISMATCH_NATIVE szErr="SUPR3LoadModule: supLoadModule returned VERR_LDR_MISMATCH_NATIVE"
00:00:02.498968 VMSetError: F:\tinderbox\win-4.3\src\VBox\VMM\VMMR3\PDMLdr.cpp(703) int __cdecl pdmR3LoadR0U(struct UVM *,const char *,const char *,const char *); rc=VERR_LDR_MISMATCH_NATIVE
00:00:02.498974 VMSetError: Cannot load R0 module C:\Program Files\Oracle\VirtualBox/VBoxDD2R0.r0: SUPR3LoadModule: supLoadModule returned VERR_LDR_MISMATCH_NATIVE
00:00:02.499003 AssertLogRel F:\tinderbox\win-4.3\src\VBox\Devices\PC\DevAPIC.cpp(2325) int __cdecl apicR3Construct(struct PDMDEVINS *,int,struct CFGMNODE *): RT_SUCCESS_NP(rc)
00:00:02.499022 VERR_MODULE_NOT_FOUND (-610) - Module not found.
00:00:02.499033 PDM: Failed to construct 'apic'/0! VERR_MODULE_NOT_FOUND (-610) - Module not found.
00:00:02.680971 ERROR [COM]: aRC=E_FAIL (0x80004005) aIID={8ab7c520-2442-4b66-8d74-4ff1e195d2b6} aComponent={Console} aText={Cannot load R0 module C:\Program Files\Oracle\VirtualBox/VBoxDD2R0.r0: SUPR3LoadModule: supLoadModule returned VERR_LDR_MISMATCH_NATIVE (VERR_LDR_MISMATCH_NATIVE)}, preserve=false
00:00:02.708761 Power up failed (vrc=VERR_MODULE_NOT_FOUND, rc=E_FAIL (0X80004005))
```

More, this does not solve my previous issue about missing npm command.
Please, any help ?

**Comment by lioman** [comment]
vagrant 1.7.4
VirtualBox: 4.3.10
Host: Ubuntu 14.04
Box: Ubuntu 14.04

I've got the same problem with npm

**Comment by tlartaud** [comment]
@lioman My temporary fix if you need

Disable nodejs auto-install
```
nodejs:
    install: '0'
    npm_packages: {  }
```

Create `puppet/files/exec-once/install-nodejs.sh`
```
#!/bin/bash

apt-get --assume-yes install software-properties-common python-software-properties python g++ make
add-apt-repository ppa:chris-lea/node.js
apt-get update
apt-get  --assume-yes install nodejs
npm install -g bower
npm install -g grunt-cli
```

**Comment by lioman** [comment]
That's how I used it before there was this option.
But instead of this PPA I would use https://nodesource.com/blog/nodejs-v012-iojs-and-the-nodesource-linux-repositories

**Comment by rulin132** [comment]
Ran into the same issue using Ubuntu 14.04 x64. Expanding on the solution from @tlartaud with the recommended source for installation given by @lioman, this is what worked for me without a hitch.

Create puphpet/files/exec-once/install-nodejs.sh

    #!/bin/bash
    
    curl -sL https://deb.nodesource.com/setup_0.12 | sudo bash -
    sudo apt-get install -y nodejs
    npm install -g bower
    npm install -g grunt-cli

**Comment by ianoshorty** [comment]
The fix referenced by [Martinsteinorth](https://github.com/martinsteinorth) in [#1886](https://github.com/puphpet/puphpet/issues/1886) worked for me.

Replace 
```shell
LATEST_NODE=$(curl 'http://nodejs.org/dist/latest/SHASUMS.txt' | grep "${FILENAME}" | awk '{ print $2 }')
```

With 
```shell
LATEST_NODE=$(curl 'http://nodejs.org/dist/latest/SHASUMS256.txt' | grep "${FILENAME}" | awk '{ print $2 }')
```
In `puphpet/puppet/modules/puphpet/templates/nodejs/install.erb`

**Comment by tlartaud** [comment]
@ianoshorty thanks for reposting this for us. Works like a charm.

**Comment by khallmark** [comment]
I am also experiencing this issue:

vagrant 1.7.4
Parallels Pro 11
Host: Mac OS X 10.10.5
Box: Ubuntu 14.04

**Comment by maltyxx** [comment]
Nodejs v0.12

LATEST_NODE=$(curl 'http://nodejs.org/dist/latest-v0.12.x/SHASUMS.txt' | grep "${FILENAME}" | awk '{ print $2 }')

**Comment by JohnHeroHD** [comment]
@ianoshorty - fix confirmed! My box works perfectly again.

**Comment by jtreminio** [comment]
Merged puphpet/puppet-puphpet/pull/26

**Comment by akireikin** [comment]
@ianoshorty thank you very much!
